packages/zoe/contracts/autoswap.js

```diff
@@ -21,303 +20,195 @@ export const makeContract = harden((zoe, terms) => {
-  });
-  return harden({
-    instance: autoswap,
-    assays,
+      rejectOffer,
+      hasValidPayoutRules,
+      vectorWith,
+      vectorWithout,
+      makeEmptyOffer,
+    } = makeHelpers(zoe, assays);
+    const getPrice = makeGetPrice(zoe, assays);
+    const getPoolUnits = () => zoe.getOffer(poolHandle).units;
+
+    return makeEmptyOffer().then(handle => {
+      poolHandle = handle;
+
+      const makeInvite = () => {
+        const seat = harden({
+          swap: () => {
+            let UNITS_IN_INDEX;
+            const unitsInFirst = ['offerAtMost', 'wantAtLeast', 'wantAtLeast'];
+            const unitsInSecond = ['wantAtLeast', 'offerAtMost', 'wantAtLeast'];
+            if (hasValidPayoutRules(unitsInFirst, inviteHandle)) {
+              UNITS_IN_INDEX = 0;
+            } else if (hasValidPayoutRules(unitsInSecond, inviteHandle)) {
+              UNITS_IN_INDEX = 1;
+            } else {
+              throw rejectOffer(inviteHandle);
+            }
+            const UNITS_OUT_INDEX = Nat(1 - UNITS_IN_INDEX);
+            const { newPoolUnitsArray, unitsOut } = getPrice(
+              getPoolUnits(),
+              zoe.getOffer(inviteHandle).units[UNITS_IN_INDEX],
+            );
+
+            const wantedUnits = zoe.getOffer(inviteHandle).units[
+              UNITS_OUT_INDEX
+            ];
+            const satisfiesWantedUnits = () =>
+              unitOpsArray[UNITS_OUT_INDEX].includes(unitsOut, wantedUnits);
+            if (!satisfiesWantedUnits()) {
+              throw rejectOffer(inviteHandle);
+            }
+
+            const newUserUnits = unitOpsArray.map(unitOps => unitOps.empty());
+            newUserUnits[UNITS_OUT_INDEX] = unitsOut;
+
+            zoe.reallocate(
+              harden([inviteHandle, poolHandle]),
+              harden([newUserUnits, newPoolUnitsArray]),
+            );
+            zoe.complete(harden([inviteHandle]));
+            return `Swap successfully completed.`;
+          },
+          addLiquidity: () => {
+            const kinds = ['offerAtMost', 'offerAtMost', 'wantAtLeast'];
+            if (!hasValidPayoutRules(kinds, inviteHandle)) {
+              throw rejectOffer(inviteHandle);
+            }
+
+            const userUnits = zoe.getOffer(inviteHandle).units;
+            const poolUnits = getPoolUnits();
+
+            // Calculate how many liquidity tokens we should be minting.
+            // Calculations are based on the extents represented by index 0.
+            // If the current supply is zero, start off by just taking the
+            // extent at index 0 and using it as the extent for the
+            // liquidity token.
+            const liquidityExtentOut =
+              liqTokenSupply > 0
+                ? floorDivide(
+                    multiply(userUnits[0].extent, liqTokenSupply),
+                    poolUnits[0].extent,
+                  )
+                : userUnits[0].extent;
```

Comment:
This calculation is specific to this bonding curve, and so should be moved to bondingCurve.js
The maker exported from bondingCurve.js should make an object with the needed calculation methods for that bonding curve.